src/analysis/bounds.rs

```diff
@@ -172,18 +172,8 @@ impl Bounds {
             Type::Fundamental(Fundamental::Filename) => Some(AsRef(None)),
             Type::Fundamental(Fundamental::OsString)=> Some(AsRef(None)),
             Type::Fundamental(Fundamental::Utf8) if *nullable => Some(Into(Some('_'), None)),
-            Type::Class(..) if !*nullable => {
-                if env.class_hierarchy.subtypes(type_id).next().is_some() {
-                    Some(IsA(None))
-                } else {
-                    None
-                }
-            }
-            Type::Class(..) => if env.class_hierarchy.subtypes(type_id).next().is_some() {
-                Some(Into(Some('_'), Some(Box::new(IsA(None)))))
-            } else {
-                Some(Into(Some('_'), None))
```

Comment:
I also stuck with this when tried to build gio:
```rust
fn write<'a, P: IsA<Cancellable> + 'a, Q: Into<Option<&'a P>>>(&self, buffer: &[u8], cancellable: Q) -> Result<isize, Error>;
...
        let result = self.0.write(buf, None).map(|size| size as usize);
//error[E0283]: type annotations required: cannot resolve `_: glib::IsA<auto::cancellable::Cancellable>`
```
This case IMHO not needed be `IsA` so maybe just leave subtypes check in nullable case?

Reply:
For cancellable it's not needed because it's a final type, yes.

But for many other types it is needed and our subtypes check is simply broken because it does not account for subtypes that are in other crates.

Also keeping that check does not really solve anything: we already have the problem in some functions that take e.g. a nullable `gtk::Widget` (because there are subtypes check sees subtypes).

I think this needs a general solution.

We could add proper detection for final types (-> no public instance/class struct) to keep the number of these cases lower, and we could generate constants like the following for non-final types:

```rust
const NoneWidget: Option<gtk::Widget> = None;
```

That's the best I can think of right now, but maybe you have some other ideas? :)

Reply:
In my case I can use `None::<&Cancellable>` but not user about your case @sdroege

Reply:
Yeah, but that's a bit ugly. Using `gio::NoneCancellable` would probably be nicer usability-wise or what do you think?

(Note that `Cancellable` would be detected as final type so this wouldn't be necessary in that case)

Reply:
I like idea for detection final types and generate Nones,
but IMHO it not need in this PR currently and can be postponed

Reply:
And please rebase to last master too

Reply:
See https://github.com/gtk-rs/gir/pull/690

Reply:
Ah actually it is needed because otherwise we have inconsistent information for `self` if we generate a trait. `Self: IsA<Foo>` but when generating the transformations we lose that information if the type has no subtypes because of that code.

It's simply wrong right now as we have inconsistent information, and would otherwise require another workaround elsewhere.

So I think this actually has to be done as part of this PR unfortunately.

Reply:
And `Cancellable` is also not a final type unfortunately, there can be subclasses of it just fine. It even has a virtual method that subclasses can override.

So doing `IsA<Cancellable>` is actually more correct here.

Reply:
Ok, this bad, currently we stay with `None::<&Cancellable>` or `NoneCancelable` it if defined.
Note: I don't see nothing wrong with first case, less dependence 😉

Reply:
I'll add some code to generate such `None` constants, that's trivial to do

Reply:
Is the `+ 'a` mandatory after the `IsA<>`?

Reply:
Yes, otherwise `P` might not live long enough (it has to live as long as `'a` from `Q: Into<Option<&'a P>>`). This is what we generated in the past already, nothing new in that code :)